spark-cassandra-connector/src/main/scala/com/datastax/spark/connector/metrics/InputMetricsUpdater.scala

```diff
@@ -0,0 +1,80 @@
+package com.datastax.spark.connector.metrics
+
+import com.codahale.metrics.Timer
+import com.datastax.driver.core.Row
+import org.apache.spark.{SparkEnv, TaskContext}
+import org.apache.spark.executor.{DataReadMethod, InputMetrics}
+import org.apache.spark.metrics.CassandraConnectorSource
+
+private[connector] trait InputMetricsUpdater {
+  def resultSetFetchTimer: Option[Timer]
+
+  def updateMetrics(row: Row): Row
+
+  def finish(): Long
+}
+
+private class DetailedInputMetricsUpdater(metrics: InputMetrics, groupSize: Int) extends InputMetricsUpdater {
+  require(groupSize > 0)
+
+  val resultSetFetchTimer = Some(CassandraConnectorSource.readMoreRowsTimer)
+
+  private val taskTimer = CassandraConnectorSource.readTaskTimer.time()
+
+  private var cnt = 0
+  private var dataLength = metrics.bytesRead
+
+  def updateMetrics(row: Row): Row = {
+    for (i <- 0 until row.getColumnDefinitions.size() if !row.isNull(i))
+      dataLength += row.getBytesUnsafe(i).remaining()
+
+    cnt += 1
+    if (cnt == groupSize)
+      update()
+    row
+  }
+
+  @inline
+  private def update(): Unit = {
+    CassandraConnectorSource.readRowMeter.mark(cnt)
+    CassandraConnectorSource.readByteMeter.mark(metrics.bytesRead - dataLength)
```

Comment:
I'm a little lost on what's happening here. The way i'm reading this I see the following happening:

datalength is set to the current number of bytes read
datalength is increased to by the size of each row
when cnt has reached groupSize:
     we update the read byteMeter by the bytesRead - (previously obtained)bytesRead+(the size of all the rows we read before) 
     then we reset datalength to (current) bytesRead